test/cpp/thread_manager/thread_manager_test.cc

```diff
@@ -0,0 +1,99 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *is % allowed in string
+ */
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include <gflags/gflags.h>
+#include <grpc++/grpc++.h>
+#include <grpc/support/log.h>
+
+#include "test/cpp/thread_manager/thread_manager_test.h"
+#include "test/cpp/util/test_config.h"
+
+using grpc::testing::ThreadManagerTest;
+
+static const int kMinPollers = 2;
+static const int kMaxPollers = 10;
+
+static const int kPollingTimeoutMsec = 10;
+static const int kDoWorkDurationMsec = 1;
+
+static const int kNumDoWorkIterations = 10;
+
+grpc::ThreadManager::WorkStatus ThreadManagerTest::PollForWork(void **tag,
+                                                               bool *ok) {
+  {
+    std::unique_lock<grpc::mutex> lock(mu_);
+    gpr_log(GPR_INFO, "PollForWork: Entered");
+  }
+
+  WorkStatus work_status = WORK_FOUND;
+  *tag = nullptr;
+  *ok = true;
+
+  // Simulate "polling for work" by sleeping for sometime
+  std::this_thread::sleep_for(std::chrono::milliseconds(kPollingTimeoutMsec));
+
+  {
+    std::unique_lock<grpc::mutex> lock(mu_);
+    num_calls_++;
+    if (num_calls_ > kNumDoWorkIterations) {
+      gpr_log(GPR_DEBUG, "PollForWork: Returning shutdown");
+      work_status = SHUTDOWN;
+      ThreadManager::Shutdown();
+    }
+  }
+
+  return work_status;
+}
+
+void ThreadManagerTest::DoWork(void *tag, bool ok) {
+  {
+    std::unique_lock<grpc::mutex> lock(mu_);
+    gpr_log(GPR_DEBUG, "DoWork()");
+  }
+
+  // Simulate "doing work" by sleeping
+  std::this_thread::sleep_for(std::chrono::milliseconds(kDoWorkDurationMsec));
+}
+
+int main(int argc, char **argv) {
+  grpc::testing::InitTest(&argc, &argv, true);
+  ThreadManagerTest test_rpc_manager(kMinPollers, kMaxPollers);
+  test_rpc_manager.Initialize();
+  test_rpc_manager.Wait();
+
+  return 0;
+}
```

Comment:
Please make this test check that the return values are as expected and also deal with cases where things didn't go as planned (e.g., timeouts).